Compute a fingerprint for constants and import_ref instructions.

zygoloid opened the pull request:
Use it in the instruction namer to make instruction names more stable
across unrelated changes to the toolchain or the prelude.

zygoloid committed:
Compute a fingerprint for constant instructions.

Use it in the instruction namer to make instruction names more stable
across unrelated changes to the toolchain or the prelude.

zygoloid committed:
Add fingerprinting for the remaining cases.

zygoloid committed:
Replace switch with table dispatch.

zygoloid committed:
Fingerprint across files.

zygoloid committed:
Also use a fingerprint for import refs.

zygoloid committed:
Factor out block handling.

zygoloid committed:
Fix fingerprinting of entity parents.

zygoloid committed:
clang-format

zygoloid commented:
This removes the vast majority of the churn from #4753, where there are now unrelated changes in only 10 files rather than ~150. Of those:
- in 5 files, the output (SemIR or LLVM IR) contains a raw instruction ID
- in 4 files, the hash printed for an instruction got longer to remain unambiguous
- in 1 file, the `@impl.N` numbering changed

We could address the second point in most cases by increasing the minimum length of hash we include, and could address the third point by also including a hash in the names of impls (and perhaps other entities too).

danakj commented on toolchain/sem_ir/inst_fingerprinter.cpp:
```diff
@@ -0,0 +1,338 @@
+};
+}  // namespace
+
+auto InstFingerprinter::GetOrCompute(const File& file, InstId inst_id)
```
Doesn't the style guide require `const File*` here because the pointer is stored beyond the scope of the function body?

zygoloid replied:
Yeah, I suppose so. (We don't perform any memory accesses through it after this function returns, but we do use it as a cache key so we require the file to outlive this object.) Changed.

danakj commented on toolchain/sem_ir/inst_fingerprinter.cpp:
```diff
@@ -0,0 +1,338 @@
+  // to build a Merkle tree containing a fingerprint for the current
+  // instruction.
+  llvm::SmallVector<llvm::stable_hash> contents = {};
+  // Known cached instruction fingerprints.
```
```suggestion
  // Known cached instruction fingerprints. Each item in `todo` will be added to the cache if not already present.
```

danakj commented on toolchain/sem_ir/inst_fingerprinter.cpp:
```diff
@@ -0,0 +1,338 @@
+  }
+
+  auto Add(IdentifierId ident_id) -> void {
+    AddString(ident_id.is_valid() ? sem_ir->identifiers().Get(ident_id) : "");
```
Why not AddInvalid() if not valid here? If there's a reason maybe a comment could explain?

zygoloid replied:
This was just an arbitrary choice. I thought I'd added this because I was hitting this case when testing, but it looks like I'm not any more (maybe I never was?) so I've just removed the invalid check.

danakj commented on toolchain/sem_ir/inst_fingerprinter.cpp:
```diff
@@ -0,0 +1,338 @@
+        Add(inst.type_id());
+      }
+
+      for (auto [arg, kind] : {std::pair(inst.arg0(), arg0_kind),
+                               std::pair(inst.arg1(), arg1_kind)}) {
+        AddWithKind(arg, kind);
+      }
```
This seems like two function calls would be a clearer way to write it, given it's just a list of two?
```suggestion
      AddWithKind(inst.arg0(), arg0_kind);
      AddWithKind(inst.arg1(), arg1_kind);
```

zygoloid replied:
Thanks! I used to have a big `switch` inside this loop but missed this cleanup when moving it out to a separate  function.

danakj commented on toolchain/sem_ir/inst_fingerprinter.cpp:
```diff
@@ -0,0 +1,338 @@
+      }
+
+      // If we didn't add any work, we have a fingerprint for this instruction.
+      if (todo.size() == init_size) {
```
If there is more work, we loop around, and we clear `contents`, so does that not end up throwing away all the previous work we just did?

zygoloid replied:
Yes. If there is more work, we can't compute the fingerprint of the current instruction yet and `contents` is incorrect, and we rebuild it from scratch when we return to this instruction.

danakj commented on toolchain/sem_ir/inst_fingerprinter.cpp:
```diff
@@ -0,0 +1,338 @@
+      if (todo.size() == init_size) {
+        auto fingerprint = llvm::stable_hash_combine(contents);
+        fingerprints->Insert(std::pair(next_sem_ir, next_inst_id), fingerprint);
+        todo.pop_back();
```
This seems to me like it's dropping one extra thing from `todo`, is it? The first thing the loop would do is pop the back, so the item currently being worked on is not in the `todo`. So this is popping the next thing to work on, and throwing it away?

zygoloid replied:
The loop doesn't pop the instruction unless it's already got a fingerprint (in which case it immediately continues to the next instruction and doesn't get here). I've added some comments to try to clarify the approach here.

danakj commented on toolchain/sem_ir/inst_namer.cpp:
```diff
@@ -245,16 +249,28 @@ auto InstNamer::Namespace::AllocateName(const InstNamer& inst_namer,
+  } else {
+    uint64_t fingerprint = std::get<uint64_t>(loc_id_or_fingerprint);
+    llvm::raw_string_ostream out(name);
+    out << ".";
```
What would you think of a prefix for fingerprints like the `.loc` for location ids? Like `.fin` or `.fp` or something?

zygoloid replied:
Hm. If we want to make them visually distinct from sequential disambiguators (even when they start with a digit), how about increasing the minimum length to, say, 3 instead? There's still a chance they'll appear to be decimal integers, but three-digit disambiguators should be vanishingly rare. Increasing the length will also remove some more churn.

I've gone ahead and done that; I think it's now relatively clear in the SemIR which parts of names are fingerprints. There's still a risk we generate a name like `%foo.bad` which might look like it means something in English. I think this strikes a reasonable balance between name length and risk of confusion, but I'd be happy to add something before the fingerprint too if you think it's worth it with the longer fingerprints.

danakj replied:
Thanks, yes I like the 3 digits a lot more.

zygoloid committed:
Apply suggestions from code review

Co-authored-by: Dana Jansens <danakj@orodu.net>

zygoloid committed:
Add some comments.

zygoloid committed:
Remove unused check for invalid identifier.

zygoloid committed:
Pass pointer.

zygoloid committed:
Increase minimum length to 3, to minimize the risk a hash will be
confused with a sequential disambiguator and to further reduce the risk
of churn.

zygoloid committed:
Increase to a minimum of 4 chars of hash.

zygoloid committed:
Merge branch 'trunk' into toolchain-hash-constants

zygoloid committed:
Revert to 3 chars. I find 3 hex digits more comfortable to read and scan for in the IR than 4.
